### PR TITLE
SI-7331, SI-6843, SI-7731, parser entry point refactoring

### PR DESCRIPTION
**Pull request by densh**
This commits fixes following bugs:
- SI-7331 tb.parse returns unpositioned trees
- SI-6843 better error reporting for quasiquotes
- SI-7731 CannotHaveAttrs has inconsistent handling of setPos and setType

And additionaly refactors parser entry points to make parsing stats consistent across repl, toolbox, script runner and quasiquotes.

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Trees.scala [hunk 1]
```diff
@@ -1017,14 +1017,16 @@ trait Trees extends api.Trees { self: SymbolTable =>
   trait CannotHaveAttrs extends Tree {
     override def canHaveAttrs = false
 
-    private def unsupported(what: String, args: Any*) =
-      throw new UnsupportedOperationException(s"$what($args) inapplicable for "+self.toString)
+    private def requireLegal(value: Any, allowed: Any, what: String) =
+      require(value == allowed, s"can't set $what for $self to value other than $allowed")
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Trees.scala:1021]: Could be unit tested. Probably fits best in a new JUnit test case under test/junit.

**Comment by retronym** [comment]
s/two logical changes/three logical changes/?

**Comment by densh** [comment]
PLS REBUILD/pr-scala@d5f4b5b

**Comment by densh** [comment]
Updated pr with junit tests for `CannotHaveAttrs` and changes to `correspondingPosition` suggested by @retronym 

I've also added assertThrows utility function that needs additional review. Probably by @gkossakowski

**Comment by densh** [comment]
Extended `CannotHaveAttrs` tests to include tests for all of the typical instances.

**Comment by adriaanm** [comment]
PLS REBUILD/pr-scala@c22d877

**Pull request closed without merging**